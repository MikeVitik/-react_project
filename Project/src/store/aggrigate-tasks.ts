import { Task } from "./slicies/tasks-slice";
import { normalizeDay } from "./utils/normalize-day";
export type FilterType = "currentWeek" | "pastWeek" | "twoWeeksAgo";
export interface StatisticValues {
  day: number;
  completedPomodoro: number;
  workTime: number;
  pauseCount: number;
  pauseTime: number;
}

export const aggrigateTasks = (tasks: Task[]): StatisticValues[] => {
  const result = Object.values(
    tasks
      .sort((a, b) => a.startDate.getTime() - b.startDate.getTime())
      .reduce<{
        [key: string]: StatisticValues;
      }>((aggrigate, task) => {
        const key = normalizeDay(task.startDate.getDay());
        if (aggrigate[key]) {
          aggrigate[key] = {
            day: key,
            completedPomodoro:
              aggrigate[key].completedPomodoro + task.completedPomodoro! / 1000,
            workTime: aggrigate[key].workTime + task.workTime! / 1000,
            pauseCount: aggrigate[key].pauseCount + task.pauseCount!,
            pauseTime: Math.floor(
              aggrigate[key].pauseTime + task.pauseTime! / 1000
            ),
          };
        } else {
          aggrigate[key] = {
            day: key,
            completedPomodoro: task.completedPomodoro! / 1000,
            workTime: task.workTime! / 1000,
            pauseCount: task.pauseCount!,
            pauseTime: Math.floor(task.pauseTime! / 1000),
          };
        }
        return aggrigate;
      }, {})
  );
  const dayRemains = 7 - result.length;
  if (dayRemains > 0) {
    const days = [...Array(dayRemains)].map((_, i) => ({
      day: i + 1,
      completedPomodoro: 0,
      workTime: 0,
      pauseCount: 0,
      pauseTime: 0,
    }));
    return [...result, ...days];
  }
  return result;
};
