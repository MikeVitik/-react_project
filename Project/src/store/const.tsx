export const BREAK_TIME = 10 * 1000;
export const TASK_TIME = 20 * 1000;
export const TIME_OF_ONE_POMODORO = 25 * 60 * 1000;

export const TOTAL_MILLISECONDS_IN_DAY = 1000 * 60 * 60 * 24;
