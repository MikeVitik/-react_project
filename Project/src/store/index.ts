import { taskEditorSlice } from "./slicies/task-editor";

import { tasksSlice } from "./slicies/tasks-slice";

import { currentTaskSlice } from "./slicies/current-task";
import { timerSlice } from "./slicies/timer-slice";

export type { CurrentTask } from "./slicies/current-task";

export const { getMode } = currentTaskSlice.selectors;

export {
  configureAppStore,
  currentTask,
  updateTime,
  useAppDispatch,
} from "./store";

export { taskActions } from "./store";

export type { RootState } from "./store";
export const {
  decrementPomodoro,
  deleteTask,
  incrementPomodoro,
  changeTask,
  addTask,
} = tasksSlice.actions;

export { createTask } from "./store";
export const { selectCanDecrement, selectTotalTime } = tasksSlice.selectors;

export const { cancelTaskEdit, editTask, updateEditTaskName } =
  taskEditorSlice.actions;
export const { getTimerValue } = timerSlice.selectors;
