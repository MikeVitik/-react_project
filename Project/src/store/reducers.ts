import { combineSlices } from "@reduxjs/toolkit";
import { currentTaskSlice } from "./slicies/current-task";
import { taskEditorSlice } from "./slicies/task-editor";
import { tasksSlice } from "./slicies/tasks-slice";
import { timerSlice } from "./slicies/timer-slice";

export const rootReducer = combineSlices(
  taskEditorSlice,
  tasksSlice,
  currentTaskSlice,
  timerSlice
);
