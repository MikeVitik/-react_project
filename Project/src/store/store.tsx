import {
  ActionCreator,
  configureStore,
  createAsyncThunk,
  createListenerMiddleware,
  createSelector,
  isAnyOf,
} from "@reduxjs/toolkit";
import { useDispatch } from "react-redux";
import { rootReducer } from "./reducers";
import { CurrentTask, currentTaskSlice } from "./slicies/current-task";
import { SerialiazableTask, addTask, tasksSlice } from "./slicies/tasks-slice";
import { timerSlice } from "./slicies/timer-slice";
export const updateTime = createAsyncThunk<void, number, { state: RootState }>(
  "updateTimer",
  (delta: number, thunkAPI) => {
    thunkAPI.dispatch(timerSlice.actions.updateTime(delta));
    if (timerSlice.selectSlice(thunkAPI.getState()).state === "end") {
      // thunkAPI.dispatch(timerEnd)
    }
  }
);

const BREAK_TIME = 10 * 1000;
const TASK_TIME = 20 * 1000;

export const createTask = createAsyncThunk<void, string, { state: RootState }>(
  "createTask",
  (text, { dispatch, getState }) => {
    const addTaskAction = addTask(text);
    dispatch(addTaskAction);
    if (currentTask(getState()).taskId === undefined) {
      dispatch(currentTaskSlice.actions.initTask(addTaskAction.payload.id));
      dispatch(timerSlice.actions.create(TASK_TIME));
    }
  }
);

export const startTask = createAsyncThunk<void, void, { state: RootState }>(
  "startTask",
  (_, { dispatch }) => {
    dispatch(currentTaskSlice.actions.startTask());
    dispatch(timerSlice.actions.start());
  }
);

export const pauseTask = createAsyncThunk<void, void, { state: RootState }>(
  "pauseTask",
  (_, { dispatch }) => {
    dispatch(currentTaskSlice.actions.pauseTask());
    dispatch(timerSlice.actions.stop());
  }
);
export const stopTask = createAsyncThunk<void, void, { state: RootState }>(
  "stopTask",
  (_, { dispatch }) => {
    dispatch(currentTaskSlice.actions.stopTask());
    dispatch(timerSlice.actions.stop());
  }
);
export const continueTask = createAsyncThunk<void, void, { state: RootState }>(
  "continueTask",
  (_, { dispatch }) => {
    dispatch(currentTaskSlice.actions.continueTask());
    dispatch(timerSlice.actions.start());
  }
);

export const startBreak = createAsyncThunk<void, void, { state: RootState }>(
  "startBreak",
  (_, { dispatch, getState }) => {
    dispatch(currentTaskSlice.actions.startBreak());
    dispatch(timerSlice.actions.create(BREAK_TIME));
  }
);

export const pauseBreak = createAsyncThunk<void, void, { state: RootState }>(
  "pauseBreak",
  (_, { dispatch, getState }) => {
    dispatch(currentTaskSlice.actions.pauseBreak());
    dispatch(timerSlice.actions.stop());
  }
);
export const continueBreak = createAsyncThunk<void, void, { state: RootState }>(
  "continueBreak",
  (_, { dispatch, getState }) => {
    dispatch(currentTaskSlice.actions.continueBreak());
    dispatch(timerSlice.actions.start());
  }
);
export const nextOrContinueTask = createAsyncThunk<
  void,
  void,
  { state: RootState }
>("nextOrContinueTask", (_, { dispatch, getState }) => {
  dispatch(currentTaskSlice.actions.nextOrContinueTask());
});
export const completeTask = createAsyncThunk<void, void, { state: RootState }>(
  "completeTask",
  (_, { dispatch, getState }) => {
    // dispatch(task ) getState()
    dispatch(currentTaskSlice.actions.pauseBreak());
    // dispatch(timerSlice.actions.stop());
  }
);

const fakeTask = { id: -1, name: "", pomodoroCount: 0 };

export const currentTask = createSelector(
  currentTaskSlice.selectSlice,
  (state: RootState) => {
    const currentTask = currentTaskSlice.selectSlice(state);
    if (currentTask.taskId !== undefined) {
      return tasksSlice.selectors.getTask(state, currentTask.taskId);
    } else {
      return fakeTask;
    }
  },
  (currentTask, task) => {
    return {
      ...currentTask,
      ...task,
    };
  }
);

const mapStateToPrimaryName: {
  [key in CurrentTask["state"]]: string;
} = {
  inited: "Старт",
  workInited: "Старт",
  workTimer: "Пауза",
  workPause: "Cтарт",
  break: "Старт",
  breakTimer: "Пауза",
  breakStop: "Продолжить",
  workStop: "Продолжить",
};

const mapStateToPrimaryAction: {
  [key in CurrentTask["state"]]: (() => ActionCreator<void>) | undefined;
} = {
  inited: undefined,
  workInited: startTask,
  workTimer: pauseTask,
  workPause: continueTask,
  break: continueBreak,
  breakTimer: pauseBreak,
  breakStop: continueBreak,
  workStop: pauseTask,
};
const mapStateToSecondaryName: {
  [key in CurrentTask["state"]]: string;
} = {
  inited: "Стоп",
  workInited: "Стоп",
  workTimer: "Стоп",
  workPause: "Стоп",
  break: "Стоп",
  breakTimer: "Продолжить",
  breakStop: "Продолжить",
  workStop: "Сделано",
};

const mapStateToSecondaryAction: {
  [key in CurrentTask["state"]]: (() => ActionCreator<void>) | undefined;
} = {
  inited: undefined,
  workInited: undefined,
  workTimer: startBreak,
  workPause: startBreak,
  break: nextOrContinueTask,
  breakTimer: nextOrContinueTask,
  breakStop: nextOrContinueTask,
  workStop: completeTask,
};

export const taskActions = createSelector(
  (state: RootState) => state.currentTask.state,
  (currentState) => {
    return {
      primaryActionName: mapStateToPrimaryName[currentState],
      primaryAction: mapStateToPrimaryAction[currentState],
      secondaryActionName: mapStateToSecondaryName[currentState],
      secondaryAction: mapStateToSecondaryAction[currentState],
    };
  }
);

export function generateTasks(
  currentDate = new Date(),
  seed = 1
): SerialiazableTask[] {
  const totalMilisecondInDay = 1000 * 60 * 60 * 24;

  function random(max: number) {
    var x = Math.sin(seed++) * 10000;
    return Math.floor(max * (x - Math.floor(x)));
  }

  const result: SerialiazableTask[] = [];
  let id = -1;
  for (let i = 0; i < 7 * 3; i++) {
    const completedPomodoro = (random(4) + 1) * TASK_TIME;
    result.push({
      name: "Generated task " + i,
      id,
      startDateString: currentDate.toUTCString(),
      isCompleted: true,
      pomodoroCount: 3,
      completedPomodoro,
      workTime: Math.floor(
        completedPomodoro + completedPomodoro * (random(30) / 100)
      ),
      pauseCount: random(4),
      pauseTime: random(20 * 1000),
    });
    currentDate = new Date(currentDate.getTime() - totalMilisecondInDay);
    id = id - 1;
  }
  return result.reverse();
}

export function configureAppStore(preloadedState = {}) {
  const listenerMiddleware = createListenerMiddleware();
  listenerMiddleware.startListening({
    predicate: isAnyOf(addTask),
    effect: () => {
      console.log("Added___");
    },
  });
  const store = configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().prepend(listenerMiddleware.middleware),
  });
  if (process.env.NODE_ENV !== "production" && (module as any).hot) {
    (module as any).hot.accept("./reducers", () =>
      store.replaceReducer(rootReducer)
    );
  }
  return store;
}

type Store = ReturnType<typeof configureAppStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = Store["dispatch"];
export const useAppDispatch = useDispatch.withTypes<AppDispatch>(); // Export a hook that can be reused to resolve types
