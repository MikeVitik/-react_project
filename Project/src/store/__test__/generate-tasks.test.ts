import { generateTasks } from "../store";

describe("generateTasks", () => {
  it("should generate task for 2 weeks", () => {
    const tasks = generateTasks(new Date(2024, 1, 20));
    expect(tasks.length).toBe(15);
    expect(tasks.filter(({ id }) => id < 0).length).toBe(15);
    expect(tasks).toMatchSnapshot();
  });
});
