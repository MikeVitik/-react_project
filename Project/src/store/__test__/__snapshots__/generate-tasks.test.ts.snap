// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTasks should generate task for 2 weeks 1`] = `
Array [
  Object {
    "completedPomodoro": 40000,
    "id": -15,
    "isCompleted": true,
    "name": "Generated task 14",
    "pauseCount": 1,
    "pauseTime": 17875,
    "pomodoroCount": 3,
    "startDateString": "Mon, 05 Feb 2024 21:00:00 GMT",
    "workTime": 145296,
  },
  Object {
    "completedPomodoro": 20000,
    "id": -14,
    "isCompleted": true,
    "name": "Generated task 13",
    "pauseCount": 1,
    "pauseTime": 9799,
    "pomodoroCount": 3,
    "startDateString": "Tue, 06 Feb 2024 21:00:00 GMT",
    "workTime": 21902,
  },
  Object {
    "completedPomodoro": 20000,
    "id": -13,
    "isCompleted": true,
    "name": "Generated task 12",
    "pauseCount": 1,
    "pauseTime": 5518,
    "pomodoroCount": 3,
    "startDateString": "Wed, 07 Feb 2024 21:00:00 GMT",
    "workTime": 50292,
  },
  Object {
    "completedPomodoro": 0,
    "id": -12,
    "isCompleted": true,
    "name": "Generated task 11",
    "pauseCount": 2,
    "pauseTime": 9067,
    "pomodoroCount": 3,
    "startDateString": "Thu, 08 Feb 2024 21:00:00 GMT",
    "workTime": 176695,
  },
  Object {
    "completedPomodoro": 60000,
    "id": -11,
    "isCompleted": true,
    "name": "Generated task 10",
    "pauseCount": 1,
    "pauseTime": 385,
    "pomodoroCount": 3,
    "startDateString": "Fri, 09 Feb 2024 21:00:00 GMT",
    "workTime": 156904,
  },
  Object {
    "completedPomodoro": 40000,
    "id": -10,
    "isCompleted": true,
    "name": "Generated task 9",
    "pauseCount": 3,
    "pauseTime": 2632,
    "pomodoroCount": 3,
    "startDateString": "Sat, 10 Feb 2024 21:00:00 GMT",
    "workTime": 137157,
  },
  Object {
    "completedPomodoro": 0,
    "id": -9,
    "isCompleted": true,
    "name": "Generated task 8",
    "pauseCount": 0,
    "pauseTime": 4229,
    "pomodoroCount": 3,
    "startDateString": "Sun, 11 Feb 2024 21:00:00 GMT",
    "workTime": 165372,
  },
  Object {
    "completedPomodoro": 40000,
    "id": -8,
    "isCompleted": true,
    "name": "Generated task 7",
    "pauseCount": 2,
    "pauseTime": 5336,
    "pomodoroCount": 3,
    "startDateString": "Mon, 12 Feb 2024 21:00:00 GMT",
    "workTime": 136751,
  },
  Object {
    "completedPomodoro": 20000,
    "id": -7,
    "isCompleted": true,
    "name": "Generated task 6",
    "pauseCount": 3,
    "pauseTime": 1157,
    "pomodoroCount": 3,
    "startDateString": "Tue, 13 Feb 2024 21:00:00 GMT",
    "workTime": 116900,
  },
  Object {
    "completedPomodoro": 40000,
    "id": -6,
    "isCompleted": true,
    "name": "Generated task 5",
    "pauseCount": 3,
    "pauseTime": 4327,
    "pomodoroCount": 3,
    "startDateString": "Wed, 14 Feb 2024 21:00:00 GMT",
    "workTime": 97381,
  },
  Object {
    "completedPomodoro": 0,
    "id": -5,
    "isCompleted": true,
    "name": "Generated task 4",
    "pauseCount": 3,
    "pauseTime": 9050,
    "pomodoroCount": 3,
    "startDateString": "Thu, 15 Feb 2024 21:00:00 GMT",
    "workTime": 25506,
  },
  Object {
    "completedPomodoro": 40000,
    "id": -4,
    "isCompleted": true,
    "name": "Generated task 3",
    "pauseCount": 3,
    "pauseTime": 19336,
    "pomodoroCount": 3,
    "startDateString": "Fri, 16 Feb 2024 21:00:00 GMT",
    "workTime": 14711,
  },
  Object {
    "completedPomodoro": 0,
    "id": -3,
    "isCompleted": true,
    "name": "Generated task 2",
    "pauseCount": 0,
    "pauseTime": 5416,
    "pomodoroCount": 3,
    "startDateString": "Sat, 17 Feb 2024 21:00:00 GMT",
    "workTime": 157778,
  },
  Object {
    "completedPomodoro": 60000,
    "id": -2,
    "isCompleted": true,
    "name": "Generated task 1",
    "pauseCount": 3,
    "pauseTime": 11649,
    "pomodoroCount": 3,
    "startDateString": "Sun, 18 Feb 2024 21:00:00 GMT",
    "workTime": 169003,
  },
  Object {
    "completedPomodoro": 40000,
    "id": -1,
    "isCompleted": true,
    "name": "Generated task 0",
    "pauseCount": 0,
    "pauseTime": 19500,
    "pomodoroCount": 3,
    "startDateString": "Mon, 19 Feb 2024 21:00:00 GMT",
    "workTime": 194853,
  },
]
`;
